stdlib/runtime/string.gr

```diff
@@ -360,38 +362,59 @@ let rec heapValueToString = (ptr, extraIndents, toplevel) => {
     },
     t when t == Tags._GRAIN_ADT_HEAP_TAG => {
       // [ <value type tag>, <module_tag>, <type_tag>, <variant_tag>, <arity>, elts ... ]
-      let variantName = getVariantName(ptr)
-
-      if (variantName == -1n) {
+      let variantPtr = getVariantMetadata(ptr)
+      if (variantPtr == -1n) {
         "<enum value>"
       } else {
+        let length = WasmI32.load(variantPtr, 12n)
+        let variantName = allocateString(length)
+        Memory.copy(variantName + 8n, variantPtr + 16n, length)
+
         let variantName = WasmI32.toGrain(variantName): String
         // Check if this is a list
         if (isListVariant(variantName)) {
           listToString(ptr, extraIndents)
         } else {
-          let variantArity = WasmI32.load(ptr, 16n)
-          if (variantArity == 0n) {
-            variantName
+          let distToRecordFields = WasmI32.load(variantPtr, 4n)
+          let isRecordVariant = distToRecordFields != 0n
+          if (isRecordVariant) {
+            let fields = variantPtr + distToRecordFields
+            let recordArity = WasmI32.load(ptr, 16n)
+            let recordVariantFields = getFieldArray(fields, recordArity)
```

Comment:
I believe we're leaking the array here (and in the regular record implementation). Could you add a `Memory.decRef` to the array after the work is done? I suppose it could happen in `recordToString`, but I think it's more obvious if it's in the same scope as the call to `getFieldArray` in both places.